runtime/src/main/scala/akka/grpc/scaladsl/Metadata.scala

```diff
@@ -18,9 +23,45 @@ import akka.util.ByteString
    *         that is a text key is used.
    */
   def getText(key: String): Option[String]
+
   /**
    * @return The binary header value for `key` if one exists, if the same key has multiple values the last occurrence
    *         that is a text key is used.
    */
   def getBinary(key: String): Option[ByteString]
+
+  /**
+   * @return The metadata as a map.
+   */
+  def asMap: MetadataMap
+}
+
+class MetadataImpl(headers: immutable.Seq[HttpHeader] = immutable.Seq.empty) extends Metadata {
+  lazy private val b64Decoder = Base64.getDecoder
+  lazy private val map: MetadataMap = {
+    // REVIEWER NOTE: modeled after akka.grpc.internal.MetadataImpl.metadataMapFromGoogleGrpcMetadata
+    var entries = Map.empty[String, List[MetadataEntry]]
+    headers.foreach { header =>
+      val key = header.name()
+      val entry =
+        if (key.endsWith("-bin")) {
+          val bytes = b64Decoder.decode(header.value())
+          BytesEntry(ByteString(bytes))
+        } else {
+          val text = header.value
+          StringEntry(text)
+        }
+      if (entries.contains(key)) {
+        entries += (key -> (entry :: entries(key)))
+      } else
+        entries += (key -> (entry :: Nil))
+    }
+    entries
+  }
```

Comment:
I wish we could do this with the mutable `Map.newBuilder` instead of creating a new map for each header entry. Not sure that is doable though.

Reply:
Ya, not with having to group by header name.

Reply:
That wouldn't help because we'd still pass a map along and update, so it would just be a different way of doing the same.